{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "19f05ae4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter FASTA file:sequence.fasta\n",
      "Enter minimum length in bp for ORFs:300\n",
      "Test1\n",
      "ORF Length: 996\n",
      "Frame 1=['GAG', 'CAA', 'GAG', 'TGA', 'AGA', 'GCG', 'AAG', 'AGA', 'ATG', 'GCT', 'CCC', 'AAG', 'GGT', 'TTA', 'ATC', 'TTT', 'TTG', 'GCT', 'GTG', 'TTA', 'TGC', 'TTC', 'TCA', 'GCA', 'CTG', 'TCA', 'CTG', 'AGT', 'CGT', 'TGT', 'CTT', 'GCG', 'GAG', 'GAT', 'AAT', 'GGA', 'CTT', 'GTT', 'ATG', 'AAC', 'TTC', 'TAC', 'AAG', 'GAA', 'TCA', 'TGC', 'CCT', 'CAG', 'GCT', 'GAA', 'GAC', 'ATC', 'ATC', 'AAA', 'GAA', 'CAA', 'GTC', 'AAG', 'CTT', 'CTC', 'TAC', 'AAG', 'CGC', 'CAC', 'AAG', 'AAC', 'ACT', 'GCT', 'TTC', 'TCC', 'TGG', 'CTC', 'AGA', 'AAC', 'ATC', 'TTC', 'CAT', 'GAC', 'TGT', 'GCT', 'GTT', 'CAG', 'AGT', 'TGT', 'GAT', 'GCT', 'TCA', 'CTG', 'TTG', 'CTG', 'GAC', 'TCC', 'ACA', 'AGA', 'AGG', 'AGC', 'TTG', 'TCT', 'GAG', 'AAG', 'GAA', 'ACA', 'GAT', 'AGA', 'AGC', 'TTT', 'GGG', 'TTG', 'AGA', 'AAT', 'TTC', 'AGG', 'TAC', 'ATT', 'GAG', 'ACC', 'ATC', 'AAA', 'GAA', 'GCT', 'TTG', 'GAA', 'AGG', 'GAA', 'TGC', 'CCA', 'GGA', 'GTT', 'GTT', 'TCC', 'TGT', 'GCT', 'GAT', 'ATC', 'CTC', 'GTT', 'CTC', 'TCT', 'GCC', 'AGA', 'GAT', 'GGC', 'ATT', 'GTT', 'TCG', 'CTA', 'GGA', 'GGT', 'CCC', 'CAT', 'ATC', 'CCT', 'CTT', 'AAA', 'ACA', 'GGA', 'AGA', 'AGG', 'GAT', 'GGT', 'AGA', 'AGG', 'AGC', 'AGA', 'GCC', 'GAT', 'GTG', 'GTA', 'GAG', 'CAG', 'TTC', 'CTC', 'CCA', 'GAC', 'CAC', 'AAT', 'GAA', 'TCC', 'ATT', 'TCT', 'GCA', 'GTT', 'CTT', 'GAC', 'AAG', 'TTT', 'GGT', 'GCC', 'ATG', 'GGA', 'ATT', 'GAC', 'ACC', 'CCC', 'GGC', 'GTA', 'GTT', 'GCA', 'TTG', 'CTT', 'GGA', 'GCA', 'CAC', 'AGT', 'GTT', 'GGT', 'CGA', 'ACC', 'CAT', 'TGT', 'GTG', 'AAG', 'TTG', 'GTG', 'CAC', 'CGT', 'TTG', 'TAC', 'CCA', 'GAG', 'ATT', 'GAT', 'CCA', 'GCT', 'CTG', 'AAC', 'CCT', 'GAC', 'CAC', 'GTC', 'CCT', 'CAC', 'ATT', 'CTG', 'AAG', 'AAG', 'TGC', 'CCT', 'GAT', 'GCC', 'ATT', 'CCA', 'GAC', 'CCT', 'AAG', 'GCC', 'GTG', 'CAG', 'TAC', 'GTG', 'AGA', 'AAC', 'GAC', 'CGT', 'GGC', 'ACC', 'CCC', 'ATG', 'ATT', 'CTA', 'GAC', 'AAC', 'AAT', 'TAC', 'TAC', 'AGA', 'AAT', 'ATA', 'TTG', 'GAC', 'AAC', 'AAG', 'GGC', 'TTG', 'TTG', 'ATA', 'GTG', 'GAT', 'CAC', 'CAA', 'CTA', 'GCC', 'AAT', 'GAC', 'AAG', 'AGG', 'ACC', 'AAG', 'CCT', 'TAT', 'GTG', 'AAG', 'AAA', 'ATG', 'GCC', 'AAG', 'AGC', 'CAG', 'GAC', 'TAT', 'TTC', 'TTC', 'AAG', 'GAG', 'TTT', 'TCT', 'AGA', 'GCC', 'ATT', 'ACT', 'TTG', 'CTC', 'TCT', 'GAG', 'AAC', 'AAC', 'CCT', 'CTC', 'ACT', 'GGC', 'ACA', 'AAG', 'GGT', 'GAG', 'ATC', 'AGA', 'AAG', 'CAG', 'TGC', 'AAT', 'GCT', 'GCC', 'AAC', 'AAG', 'CAC', 'CAT', 'GAG', 'GAG', 'CCT', 'TAA', 'TTG', 'CTT', 'CCC', 'GCT', 'TAA', 'TTT', 'GGG', 'CCT', 'TGA', 'ATT', 'TTC', 'TTC', 'CCC', 'TTC', 'TCT', 'ATG', 'TGG', 'AAG', 'AAA', 'TCT', 'GTA', 'AGA', 'TAT', 'TAT', 'GCA', 'AAA', 'AAT', 'AAT', 'TAA', 'GGT', 'GTT', 'TTT', 'CTT', 'TAA', 'ATG', 'GGT', 'TGG', 'TTG', 'ATT', 'GGT', 'TCA', 'ATG', 'AAC', 'CGA', 'TCA', 'AGA', 'CCA', 'CAG', 'CAG', 'GTT', 'CAT', 'GGG', 'GAT', 'GCG', 'AGG', 'ATT', 'AAG', 'ACG', 'CTT', 'TGT', 'TTT', 'TTA', 'ATC', 'TTC', 'CGA', 'TGT', 'CAC', 'TCT', 'TGT', 'TTG', 'TTA', 'GTT', 'TGT', 'TTT', 'TTT', 'ATT', 'TTT', 'TAT', 'TTC', 'AAT', 'AAG', 'TAC', 'TGT', 'GCA', 'AGT', 'AGG', 'TTA', 'GAG', 'TTG', 'GGT', 'AGA', 'AGG', 'GCA', 'TGT', 'TCA', 'TGG', 'TGT', 'TAA', 'TTA', 'CTA', 'TGT', 'TAT', 'GTA', 'TGC', 'ATG', 'TGA', 'GTG', 'CTG', 'CTA', 'TCG', 'ATG', 'GCA', 'AGA', 'TGT', 'CAA', 'TGT', 'ATG', 'CGT', 'GTA', 'GTG', 'CTG', 'TTA', 'TCG', 'ATG', 'AGA', 'GTG', 'AAA', 'ATG', 'TTT', 'ATG', 'ATA', 'TCC', 'ACA', 'CTA', 'ATA', 'AAG', 'CTA', 'GCT', 'TGC', 'TCT', 'TGC', 'TAC', 'ATA', 'ATA', 'AAT', 'AAA', 'TCA', 'TGG', 'CCC', 'ACG', 'GTC', 'ATT', 'ATA', 'CAA', 'AAA', 'AAA', 'AAA', 'AAA', 'AAA']\n",
      "SC IDX: 8, TAA\n",
      "EC IDX: 339, TAA\n",
      "ORF Length: 324\n",
      "Frame 6=['TTT', 'TTT', 'TTT', 'TTT', 'TTT', 'TGT', 'ATA', 'ATG', 'ACC', 'GTG', 'GGC', 'CAT', 'GAT', 'TTA', 'TTT', 'ATT', 'ATG', 'TAG', 'CAA', 'GAG', 'CAA', 'GCT', 'AGC', 'TTT', 'ATT', 'AGT', 'GTG', 'GAT', 'ATC', 'ATA', 'AAC', 'ATT', 'TTC', 'ACT', 'CTC', 'ATC', 'GAT', 'AAC', 'AGC', 'ACT', 'ACA', 'CGC', 'ATA', 'CAT', 'TGA', 'CAT', 'CTT', 'GCC', 'ATC', 'GAT', 'AGC', 'AGC', 'ACT', 'CAC', 'ATG', 'CAT', 'ACA', 'TAA', 'CAT', 'AGT', 'AAT', 'TAA', 'CAC', 'CAT', 'GAA', 'CAT', 'GCC', 'CTT', 'CTA', 'CCC', 'AAC', 'TCT', 'AAC', 'CTA', 'CTT', 'GCA', 'CAG', 'TAC', 'TTA', 'TTG', 'AAA', 'TAA', 'AAA', 'ATA', 'AAA', 'AAA', 'CAA', 'ACT', 'AAC', 'AAA', 'CAA', 'GAG', 'TGA', 'CAT', 'CGG', 'AAG', 'ATT', 'AAA', 'AAA', 'CAA', 'AGC', 'GTC', 'TTA', 'ATC', 'CTC', 'GCA', 'TCC', 'CCA', 'TGA', 'ACC', 'TGC', 'TGT', 'GGT', 'CTT', 'GAT', 'CGG', 'TTC', 'ATT', 'GAA', 'CCA', 'ATC', 'AAC', 'CAA', 'CCC', 'ATT', 'TAA', 'AGA', 'AAA', 'ACA', 'CCT', 'TAA', 'TTA', 'TTT', 'TTT', 'GCA', 'TAA', 'TAT', 'CTT', 'ACA', 'GAT', 'TTC', 'TTC', 'CAC', 'ATA', 'GAG', 'AAG', 'GGG', 'AAG', 'AAA', 'ATT', 'CAA', 'GGC', 'CCA', 'AAT', 'TAA', 'GCG', 'GGA', 'AGC', 'AAT', 'TAA', 'GGC', 'TCC', 'TCA', 'TGG', 'TGC', 'TTG', 'TTG', 'GCA', 'GCA', 'TTG', 'CAC', 'TGC', 'TTT', 'CTG', 'ATC', 'TCA', 'CCC', 'TTT', 'GTG', 'CCA', 'GTG', 'AGA', 'GGG', 'TTG', 'TTC', 'TCA', 'GAG', 'AGC', 'AAA', 'GTA', 'ATG', 'GCT', 'CTA', 'GAA', 'AAC', 'TCC', 'TTG', 'AAG', 'AAA', 'TAG', 'TCC', 'TGG', 'CTC', 'TTG', 'GCC', 'ATT', 'TTC', 'TTC', 'ACA', 'TAA', 'GGC', 'TTG', 'GTC', 'CTC', 'TTG', 'TCA', 'TTG', 'GCT', 'AGT', 'TGG', 'TGA', 'TCC', 'ACT', 'ATC', 'AAC', 'AAG', 'CCC', 'TTG', 'TTG', 'TCC', 'AAT', 'ATA', 'TTT', 'CTG', 'TAG', 'TAA', 'TTG', 'TTG', 'TCT', 'AGA', 'ATC', 'ATG', 'GGG', 'GTG', 'CCA', 'CGG', 'TCG', 'TTT', 'CTC', 'ACG', 'TAC', 'TGC', 'ACG', 'GCC', 'TTA', 'GGG', 'TCT', 'GGA', 'ATG', 'GCA', 'TCA', 'GGG', 'CAC', 'TTC', 'TTC', 'AGA', 'ATG', 'TGA', 'GGG', 'ACG', 'TGG', 'TCA', 'GGG', 'TTC', 'AGA', 'GCT', 'GGA', 'TCA', 'ATC', 'TCT', 'GGG', 'TAC', 'AAA', 'CGG', 'TGC', 'ACC', 'AAC', 'TTC', 'ACA', 'CAA', 'TGG', 'GTT', 'CGA', 'CCA', 'ACA', 'CTG', 'TGT', 'GCT', 'CCA', 'AGC', 'AAT', 'GCA', 'ACT', 'ACG', 'CCG', 'GGG', 'GTG', 'TCA', 'ATT', 'CCC', 'ATG', 'GCA', 'CCA', 'AAC', 'TTG', 'TCA', 'AGA', 'ACT', 'GCA', 'GAA', 'ATG', 'GAT', 'TCA', 'TTG', 'TGG', 'TCT', 'GGG', 'AGG', 'AAC', 'TGC', 'TCT', 'ACC', 'ACA', 'TCG', 'GCT', 'CTG', 'CTC', 'CTT', 'CTA', 'CCA', 'TCC', 'CTT', 'CTT', 'CCT', 'GTT', 'TTA', 'AGA', 'GGG', 'ATA', 'TGG', 'GGA', 'CCT', 'CCT', 'AGC', 'GAA', 'ACA', 'ATG', 'CCA', 'TCT', 'CTG', 'GCA', 'GAG', 'AGA', 'ACG', 'AGG', 'ATA', 'TCA', 'GCA', 'CAG', 'GAA', 'ACA', 'ACT', 'CCT', 'GGG', 'CAT', 'TCC', 'CTT', 'TCC', 'AAA', 'GCT', 'TCT', 'TTG', 'ATG', 'GTC', 'TCA', 'ATG', 'TAC', 'CTG', 'AAA', 'TTT', 'CTC', 'AAC', 'CCA', 'AAG', 'CTT', 'CTA', 'TCT', 'GTT', 'TCC', 'TTC', 'TCA', 'GAC', 'AAG', 'CTC', 'CTT', 'CTT', 'GTG', 'GAG', 'TCC', 'AGC', 'AAC', 'AGT', 'GAA', 'GCA', 'TCA', 'CAA', 'CTC', 'TGA', 'ACA', 'GCA', 'CAG', 'TCA', 'TGG', 'AAG', 'ATG', 'TTT', 'CTG', 'AGC', 'CAG', 'GAG', 'AAA', 'GCA', 'GTG', 'TTC', 'TTG', 'TGG', 'CGC', 'TTG', 'TAG', 'AGA', 'AGC', 'TTG', 'ACT', 'TGT', 'TCT', 'TTG', 'ATG', 'ATG', 'TCT', 'TCA', 'GCC', 'TGA', 'GGG', 'CAT', 'GAT', 'TCC', 'TTG', 'TAG', 'AAG', 'TTC', 'ATA', 'ACA', 'AGT', 'CCA', 'TTA', 'TCC', 'TCC', 'GCA', 'AGA', 'CAA', 'CGA', 'CTC', 'AGT', 'GAC', 'AGT', 'GCT', 'GAG', 'AAG', 'CAT', 'AAC', 'ACA', 'GCC', 'AAA', 'AAG', 'ATT', 'AAA', 'CCC', 'TTG', 'GGA', 'GCC', 'ATT', 'CTC', 'TTC', 'GCT', 'CTT', 'CAC', 'TCT', 'TGC']\n",
      "SC IDX: 310, TGA\n",
      "EC IDX: 417, TGA\n"
     ]
    }
   ],
   "source": [
    "#get_seq_from_file function reads in a FASTA file and returns a list of tuples, where each tuple contains the header and DNA \n",
    "#sequence for a single sequence in the file\n",
    "\n",
    "def get_seq_from_file(fname):\n",
    "    with open(fname, 'r') as file:\n",
    "\n",
    "        sequences = []  # An empty list to hold the headers and DNA sequences in the FASTA file\n",
    "        header = ''# A variable to hold the current header\n",
    "        dna = '' # A variable to hold the current DNA sequence\n",
    "\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'): # If the line starts with '>'\n",
    "                if dna:\n",
    "                    sequences.append((header, dna))\n",
    "                    dna = ''\n",
    "                header = line[1:]\n",
    "            else:\n",
    "                dna += line.upper()\n",
    "        if dna:\n",
    "            sequences.append((header, dna))\n",
    "    return sequences\n",
    "\n",
    "#function takes a DNA sequence as input and returns a list of codons, which are substrings of the DNA sequence of length 3. \n",
    "\n",
    "def get_codons(seq_frame):\n",
    "    codons = []\n",
    "    for i in range(0, len(seq_frame), 3):\n",
    "        if len(seq_frame[i:i+3]) == 3:\n",
    "            codons.append(seq_frame[i:i+3])\n",
    "    return codons\n",
    "\n",
    "#function takes a DNA sequence as input and returns its reverse complement\n",
    "def reverse_complement(seq: str) -> str:    #Shatakshi Shewale\n",
    "    # Define a dictionary that maps each nucleotide base to its complement base\n",
    "    complement_dict = {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\"}   \n",
    "    \n",
    "    # Create a new string by iterating over the reversed input sequence and looking up each base in the complement dictionary\n",
    "    reverse_complement_seq = \"\".join(complement_dict.get(base, \"\") for base in reversed(seq))\n",
    "    \n",
    "    # Return the reverse complement sequence\n",
    "    return reverse_complement_seq    \n",
    "\n",
    "#function takes a DNA sequence as input and returns six possible codon frames\n",
    "def get_seq_frames(dna):    #Shatakshi Shewale\n",
    "    # Reverse complement the DNA sequence\n",
    "    dna_rev_comp = reverse_complement(dna)\n",
    "\n",
    "    # Create an empty list to store the different reading frames\n",
    "    seq_frames = []\n",
    "\n",
    "    # Add the original DNA sequence as the first reading frame\n",
    "    seq_frames.append(dna)\n",
    "\n",
    "    # Add the second reading frame, starting from the second nucleotide\n",
    "    seq_frames.append(dna[1:])\n",
    "\n",
    "    # Add the third reading frame, starting from the third nucleotide\n",
    "    seq_frames.append(dna[2:])\n",
    "\n",
    "    # Add the reverse complement of the DNA sequence as the fourth reading frame\n",
    "    seq_frames.append(dna_rev_comp)\n",
    "\n",
    "    # Add the fifth reading frame of the reverse complement, starting from the second nucleotide\n",
    "    seq_frames.append(dna_rev_comp[1:])\n",
    "\n",
    "    # Add the sixth reading frame of the reverse complement, starting from the third nucleotide\n",
    "    seq_frames.append(dna_rev_comp[2:])\n",
    "\n",
    "    # Return the list of reading frames\n",
    "    return seq_frames\n",
    "\n",
    "\n",
    "# function takes a list of codon frames and a minimum length and identifies all possible ORFs in the \n",
    "#codon frames that are longer than the minimum length\n",
    "\n",
    "def get_orfs(codon_frames_all, min_length):\n",
    "    frame_orfs = [] # stores information about identified ORFs in each frame\n",
    "    orfs = [] #stores the actual ORF sequences\n",
    "    \n",
    "    #Iteration over each codon frame\n",
    "    for fi, codon_frame in enumerate(codon_frames_all):\n",
    "        start_codon = 'ATG'\n",
    "        end_codons = ['TAA', 'TAG', 'TGA']\n",
    "        \n",
    "        sc_flag = False # initialize start codon flag to False\n",
    "        ec_flag = False # initialize end codon flag to False \n",
    "        sc_idx = None # initialize start codon index to None\n",
    "        ec_idx = None # initialize end codon index to None\n",
    "        orf_flag = False # initialize ORF flag to False\n",
    "        \n",
    "    #     print(codon_frame)\n",
    "    \n",
    "       # iterate over each codon in the current frame\n",
    "        for idx,codon in enumerate(codon_frame):\n",
    "            # if an ORF has not yet been identified\n",
    "            if not orf_flag:\n",
    "                # if a start codon has not yet been identified\n",
    "                if not sc_flag:\n",
    "                    # if the current codon is a start codon\n",
    "                    if codon == start_codon:\n",
    "                        sc_idx = idx\n",
    "                        sc_flag = True\n",
    "        #                 print(f'SC IDX: {sc_idx}, {codon}')\n",
    "                else:\n",
    "                    for i in range(sc_idx+1, len(codon_frame)):\n",
    "                        codon = codon_frame[i]\n",
    "                        if codon in end_codons:\n",
    "                            ec_idx = i\n",
    "                            ec_flag = True\n",
    "        #                     print(f'EC IDX: {ec_idx}, {codon}')\n",
    "\n",
    "                        if sc_flag and ec_flag:\n",
    "                            orf_len = (ec_idx+1-sc_idx)*3\n",
    "                            if orf_len >= min_length:\n",
    "                                print(f'ORF Length: {orf_len}')\n",
    "                                print(f'Frame {fi+1}={codon_frame}')\n",
    "                                print(f'SC IDX: {sc_idx}, {codon}')\n",
    "                                print(f'EC IDX: {ec_idx}, {codon}')\n",
    "                                orf = codon_frame[sc_idx:ec_idx+1]\n",
    "                                orfs.append(orf)\n",
    "                                frame_orfs.append((fi+1, sc_idx, orf_len, orf))\n",
    "                                orf_flag=True\n",
    "                            sc_flag = False\n",
    "                            ec_flag = False\n",
    "                            sc_idx = None\n",
    "                            ec_idx = None\n",
    "                    sc_flag = False\n",
    "                    ec_flag = False\n",
    "                    sc_idx = None\n",
    "                    ec_idx = None\n",
    "                    \n",
    "    return frame_orfs\n",
    "\n",
    "\n",
    "def write_output(header, frame_orfs):\n",
    "    for frame_orf in frame_orfs:\n",
    "        fi = frame_orf[0]\n",
    "        sc_idx = frame_orf[1]\n",
    "        orf_len = frame_orf[2]\n",
    "        orf = frame_orf[3]\n",
    "        if fi < 3:\n",
    "            pos = sc_idx*3+1\n",
    "        else:\n",
    "            pos = -(sc_idx+1)*3\n",
    "        print(f'>{header} | FRAME = {fi} POS = {pos} LEN = {orf_len}')\n",
    "        for i in range(0,len(orf),15):\n",
    "            print(' '.join(orf[i:i+15]))\n",
    "            \n",
    "            \n",
    "def run_all():\n",
    "    fname = input('Enter FASTA file:')\n",
    "    min_len = int(input('Enter minimum length in bp for ORFs:'))\n",
    "    \n",
    "    sequences = get_seq_from_file(fname)\n",
    "    with open('code_output.txt', 'w') as f:\n",
    "        f.write('This program will find ORFs for sequences. When prompted, please enter a FASTA file with 1 or more sequences.')\n",
    "        f.write('\\n\\n')\n",
    "        f.write(f'Enter FASTA file: {fname}\\n\\n')\n",
    "        f.write(f'Enter minimum length in bp for ORFs: {min_len}\\n')\n",
    "        \n",
    "        \n",
    "        for sequence in sequences:\n",
    "            header = sequence[0]\n",
    "            dna = sequence[1]\n",
    "            print(header)\n",
    "\n",
    "            seq_frames = get_seq_frames(dna)\n",
    "            codon_frames = [get_codons(seq_frame) for seq_frame in seq_frames]\n",
    "            frame_orfs = get_orfs(codon_frames, min_len)\n",
    "\n",
    "            for frame_orf in frame_orfs:\n",
    "                fi = frame_orf[0]\n",
    "                sc_idx = frame_orf[1]\n",
    "                orf_len = frame_orf[2]\n",
    "                orf = frame_orf[3]\n",
    "                if fi < 3:\n",
    "                    pos = sc_idx*3+1\n",
    "                else:\n",
    "                    pos = -(sc_idx+1)*3\n",
    "                f.write(f'>{header} | FRAME = {fi} POS = {pos} LEN = {orf_len}')\n",
    "                f.write('\\n')\n",
    "                for i in range(0,len(orf),15):\n",
    "                    f.write(' '.join(orf[i:i+15]))\n",
    "                    f.write('\\n')\n",
    "\n",
    "run_all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "437f8133",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
